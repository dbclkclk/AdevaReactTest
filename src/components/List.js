import React from "react";
import NotesStore from "../stores/notesstore";
import NoteAction from "../actions/note";
import { FontAwesomeIcon } from '@fortawesome/react-fontawesome';
import { Button, Modal, ModalHeader, ModalBody, ModalFooter, Input } from 'reactstrap';

export default class List extends React.Component {
  constructor(props) {
    super(props);
    this.state = {
        notes: [],
        note: null,
        search: null
    };
    this.renderNotes = this.renderNotes.bind(this);
    this.seeNote = this.seeNote.bind(this);
    this.close = this.close.bind(this);
    this.deleteNote = this.deleteNote.bind(this);
    this.handleState = this.handleState.bind(this);
    this.search = this.search.bind(this);
    this.editNote = this.editNote.bind(this);
  }
  componentWillMount() {
      NotesStore.addChangeListener(this.renderNotes);
      this.renderNotes();
  }
  editNote(key, note) {
      NoteAction.Edit(key, note);
  }
  renderNotes () {
      let values = NotesStore.getState();
      values = (values != null && values != undefined) ? values : [];
      this.setState({notes: values});
  }
  seeNote(note) {
     this.setState({note: note});
  }
  deleteNote(key) {
      NoteAction.Delete(key);
  }
  close () {
      this.setState({note: null});
  }
  handleState(e) {
    let search = e.target.value;
    this.setState({search: search});
  }
  search (note) {
       let result = true;
       let that = this;
       if (this.state.search != null && this.state.search != "")
            result = note.indexOf(that.state.search) > -1;
       return result;
  }
  render() {
    return (
        <div className="Container-fluid">
            <h3>Notes</h3>
            <Input type="text" name="search" id="search" onChange={this.handleState} placeholder="Enter your search term" />
            <ul className="list-group">
                    {this.state.notes.map((note, key) => 
                             (this.search(note) ? (<li className="list-group-item" key={key}><span>{note.substring(0, 10)+"..."}</span><span className="float-right"><a href="#" onClick={() => this.editNote(key, note)}><FontAwesomeIcon icon="edit" /></a>&nbsp; &nbsp;<a href="#" onClick={() => this.deleteNote(key)}><FontAwesomeIcon icon="trash" /></a>&nbsp; &nbsp;<a href="#" onClick={() => this.seeNote(note)}><FontAwesomeIcon icon="binoculars" /></a></span></li>) : "")

                    )}
            </ul>
            <Modal isOpen={this.state.note !== null}>
                <ModalHeader>
                    Note Display
                </ModalHeader>
                <ModalBody>
                    <p>{this.state.note}</p>
                </ModalBody>
                <ModalFooter>
                    <Button color="secondary" onClick={this.close}>close</Button>
                </ModalFooter>
            </Modal>
        </div>
      );
  }
}