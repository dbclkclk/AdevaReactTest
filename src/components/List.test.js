import React from "react";
import { shallow, mount } from 'enzyme';
import List from "./List";
import NotesStore from "../stores/notesstore";
import NoteAction from "../actions/note";
import {chaiexpect} from "../../tests/ulils";
import { library } from '@fortawesome/fontawesome-svg-core';
import { FontAwesomeIcon } from '@fortawesome/react-fontawesome';
import {faTrash, faBinoculars, faEdit} from '@fortawesome/free-solid-svg-icons';

let mockaddChangeListener = jest.fn();
let mockGetState = jest.fn();
let mockDelete = jest.fn();
let mockEdit = jest.fn();
NotesStore.addChangeListener = (fn) => mockaddChangeListener = fn;
NotesStore.getState = mockGetState;
NoteAction.Delete = mockDelete;
NoteAction.Edit = mockEdit;

describe("List Component", () => {
  it("it should render initial dom", () => {
    const wrapper = shallow(<List />);
    expect(wrapper.getElements()).toMatchSnapshot();
  });
  it("it should fetch the initial notes and display it", () => {
    mockGetState.mockReturnValue(["tttt", "dddd", "ggggg"]);
    const wrapper = mount(<List />);
    wrapper.update();
    chaiexpect(wrapper.find("li")).to.have.length(3);
  });
  it("if I should search, for `tttt`, it should only display one match", () => {
    mockGetState.mockReturnValue(["tttt", "dddd", "ggggg"]);
    const wrapper = mount(<List />);
    wrapper.find("#search").at(0).simulate('change', {target: {value: 'tttt', name: 'search'}});
    wrapper.update();
    chaiexpect(wrapper.find("li")).to.have.length(1);
  });
  it("Clicking on view should open modal", () => {
    mockGetState.mockReturnValue(["tttt", "dddd", "ggggg"]);
    const wrapper = mount(<List />);
    wrapper.find("li").at(0).find("a").at(2).simulate("click");
    wrapper.update();
    expect(wrapper.find('.modal-dialog').exists()).toEqual(true);
  });
  it("When click delete, should call delete", () => {
    mockGetState.mockReturnValue(["tttt", "dddd", "ggggg"]);
    const wrapper = mount(<List />);
    wrapper.find("li").at(0).find("a").at(1).simulate("click");
    wrapper.update();
    expect(mockDelete).toHaveBeenCalled();
  });
  it("When click edit, should call edit", () => {
    mockGetState.mockReturnValue(["tttt", "dddd", "ggggg"]);
    const wrapper = mount(<List />);
    wrapper.find("li").at(0).find("a").at(0).simulate("click");
    wrapper.update();
    expect(mockEdit).toHaveBeenCalled();
  });
});