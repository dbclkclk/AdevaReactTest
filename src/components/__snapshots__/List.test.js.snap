// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`List Component it should render initial dom 1`] = `
Array [
  <div
    className="Container-fluid"
  >
    <h3>
      Notes
    </h3>
    <Input
      id="search"
      name="search"
      onChange={[Function]}
      placeholder="Enter your search term"
      type="text"
    />
    <ul
      className="list-group"
    />
    <Modal
      autoFocus={true}
      backdrop={true}
      backdropTransition={
        Object {
          "mountOnEnter": true,
          "timeout": 150,
        }
      }
      centered={false}
      fade={true}
      isOpen={false}
      keyboard={true}
      modalTransition={
        Object {
          "timeout": 300,
        }
      }
      onClosed={[Function]}
      onOpened={[Function]}
      returnFocusAfterClose={true}
      role="dialog"
      scrollable={false}
      unmountOnClose={true}
      zIndex={1050}
    >
      <ModalHeader
        charCode={215}
        closeAriaLabel="Close"
        tag="h5"
        wrapTag="div"
      >
        Note Display
      </ModalHeader>
      <ModalBody
        tag="div"
      >
        <p />
      </ModalBody>
      <ModalFooter
        tag="div"
      >
        <Button
          color="secondary"
          onClick={[Function]}
          tag="button"
        >
          close
        </Button>
      </ModalFooter>
    </Modal>
  </div>,
]
`;
